src/Microsoft.Graph/Tasks/PageIterator.cs

```diff
@@ -0,0 +1,238 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+/**
+
+Spec
+    https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/tasks/PageIteratorTask.md
+**/
+
+namespace Microsoft.Graph
+{
+    /// <summary>
+    /// Use PageIterator&lt;T&gt; to automatically page through result sets across multiple calls 
+    /// and process each item in the result set.
+    /// </summary>
+    /// <typeparam name="T">The common entity type returned in the result set.</typeparam>
+    public class PageIterator<T> where T : Entity
+    {
+        private GraphServiceClient client;
+        private ICollectionPage<T> currentPage;
+        private Queue<T> pageItemQueue;
+        private Func<T, bool> processPageItemCallback;
+        
+        /// <summary>
+        /// The @odata.deltaLink returned from a delta query.
+        /// </summary>
+        public string Deltalink { get; private set; }
+        /// <summary>
+        /// The @odata.nextLink returned in a paged result.
+        /// </summary>
+        public string Nextlink { get; private set; }
+        /// <summary>
+        /// The PageIterator state.
+        /// </summary>
+        public PagingState State { get; set; }
+
+        /// <summary>
+        /// Creates the PageIterator with the results of an initial paged request. 
+        /// </summary>
+        /// <param name="client">The GraphServiceClient object used to create the NextPageRequest for a delta query.</param>
+        /// <param name="page">A generated implementation of ICollectionPage.</param>
+        /// <param name="callback">A Func delegate that processes type T in the result set and should return false if the iterator should cancel processing.</param>
+        /// <returns>A PageIterator&lt;T&gt; that will process additional result pages based on the rules specified in Func&lt;T,bool&gt; processPageItems</returns>
+        public static PageIterator<T> CreatePageIterator(GraphServiceClient client, ICollectionPage<T> page, Func<T,bool> callback)
```

Comment:
GraphServiceClient is only needed for setting up the NextPageRequest for delta queries. I now think that I should have another method called CreateDeltaPageIterator

Reply:
@MIchaelMainer I think it's good the way it is.